ffshot.c

```diff
@@ -118,24 +105,19 @@ int main(int argc, char* argv[]) {
 	*(uint32_t*)(buf + 4) = htobe32(height);
 	bwrite(buf, 8);
 
-	size_t end = width * height;
-	uint16_t r, g, b, a;
 	uint32_t i;
-	uint32_t mask = (2 << (bpc - 1)) - 1;
-	uint32_t scale = (1 << (16 - bpc)) + 1;
-	for (i=0; i < end; i++) {
+	uint32_t p;
+	uint32_t mask = (1 << bpc) - 1;
+	uint32_t scale = mask + 2;
+	for (i = 0; i < width * height; i++) {
 		// write out pixel
 		// BGRA? thanks Xorg.
-		b = (data[i] >> (0*bpc)) & mask;
-		g = (data[i] >> (1*bpc)) & mask;
-		r = (data[i] >> (2*bpc)) & mask;
-		a = (data[i] >> (3*bpc)) & mask;
-
-		uint32_t p = i * 4;
-		img[p + 0] = htobe16(r * scale);
-		img[p + 1] = htobe16(g * scale);
-		img[p + 2] = htobe16(b * scale);
-		img[p + 3] = hasa ? htobe16(a * scale) : 0xFFFF;
+
+		p = i * 4;
+		img[p + 0] = htobe16(CD(2) * scale);                             // r
```

Comment:
This code has gotten a lot harder to read.
Since the order of the color channels (Xorg BGRA and FF RGBA) differs, it is not entirely clear what is what.